### PR TITLE
switch the server to using threadgroups

### PR DESCRIPTION
**Pull request by DavidVorick**
The client needs to be switched over as well.

It got late so I didn't have time to go through and carefully review all the changes. There are probably other places in the server code where patters like `time.Sleep` or `time.After` are used, and there are probably places where goroutines are being created without being wrapped by `tg.Launch()`, all of which need to be fixed.

But I needed a nap and I wasn't sure how long I would be asleep, I wanted these changes pushed so they wouldn't be blocking anyone. Here's what the first draft looks like, probably more to come.

### CODE REVIEW

File: server/report_listener_udp.go [hunk 1]
```diff
@@ -156,40 +156,44 @@ func (server *GCAServer) managedHandleEquipmentReport(rawData []byte) {
...
-		default:
-			// Wait for the next incoming packet
 		}
 
 		// Read from the UDP socket
 		buffer := make([]byte, equipmentReportSize)
-		readBytes, _, err := server.udpConn.ReadFromUDP(buffer)
+		readBytes, _, err := udpConn.ReadFromUDP(buffer)
 		if err != nil {
 			server.logger.Error("Failed to read from UDP socket: ", err)
```
Comment (mikew686) [inline server/report_listener_udp.go:198]: Should test IsStopped() here and not log a spurious error.

File: server/report_listener_udp.go [hunk 2]
```diff
@@ -156,40 +156,44 @@ func (server *GCAServer) managedHandleEquipmentReport(rawData []byte) {
...
+		if server.tg.IsStopped() {
 			return
-		default:
-			// Wait for the next incoming packet
 		}
 
 		// Read from the UDP socket
 		buffer := make([]byte, equipmentReportSize)
-		readBytes, _, err := server.udpConn.ReadFromUDP(buffer)
+		readBytes, _, err := udpConn.ReadFromUDP(buffer)
```
Comment (mikew686) [inline server/report_listener_udp.go:196]: Test for isStopped to avoid error log

File: server/sync_listener_tcp.go [hunk 3]
```diff
@@ -60,7 +64,13 @@ func (gcas *GCAServer) threadedListenForSyncRequests(tcpReady chan struct{}) {
 		}
```
Comment (mikew686) [inline server/sync_listener_tcp.go:64]: Accept will log a spurious error on shutdown

**Comment by mikew686** [comment]
I'll work on the client part

**Comment by DavidVorick** [comment]
This should be ready to go

**Review by mikew686 — CHANGES_REQUESTED** [review]

File: server/server.go [hunk 4]
```diff
@@ -132,15 +122,36 @@ func NewGCAServer(baseDir string) (*GCAServer, error) {
...
+	server.logger = logger
+
+	// Create the http server and provision its shutdown.
+	server.mux = http.NewServeMux()
+	server.httpServer = &http.Server{
+		Addr: serverIP + ":" + strconv.Itoa(httpPort),
+		Handler: server.mux,
+	}
+	server.tg.OnStop(func() error {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
[the remark this answers is not in the document]
Comment (DavidVorick) [reply server/server.go:145]: I fixed this in a follow-up PR

**Pull request merged**